lib/spack/docs/tutorial_buildsystems.rst

```diff
@@ -0,0 +1,498 @@
+each build system and give examples on how these can be manipulated to 
+install a package.
+
+-----------------------
+Package class Hierarchy
+-----------------------
+.. graphviz::
+
+    digraph {
+        {Package, MakefilePackage, AutotoolsPackage, CMakePackage, PythonPackage, RPackage} -> PackageBaseClass
+    }
+
+The above diagram gives a high level view of the class hierarchy and how each
+package relates. Each subclass inherits from the :code:`PackageBaseClass` 
+super class. The bulk of the work is done in this super class which includes 
+fetching, extracting to a staging directory and installing. Each subclass 
+thus adds additional functionality that includes adding the correct parameters 
+for the build system. In the following sections, we will go over examples of 
+how to utilize each subclass and to see how powerful these abstractions are 
+when packaging.
+
+-----------------
+Package
+-----------------
+
+We've already seen examples of a Package class in our walkthrough for writing
+package files, so we won't be spending much time with them here. Briefly,
+the Package class allows for abitrary control over the build process, whereas
+subclasses rely on certain patterns (e.g. :code:`configure` :code:`make` 
+:code:`make install`) to be useful. Package classes are particularly useful 
+for packages that have a non-conventional way of being built since the packager 
+can utilize some of Spack's helper functions to customize the building and 
+installing of a package.
+
+-------------------
+Autotools
+-------------------
+
+As we have seen earlier, packages using :code:`Autotools` use :code:`configure`, 
+:code:`make` and :code:`make install` commands to execute the build and 
+install process. In our :code:`Package` class, your typical build incantation will 
+consist of the following:
+
+.. code-block:: python
+
+    def install(self, spec, prefix):
+        configure("--prefix=" + prefix)
+        make()
+        make("install")
+
+You'll see that this looks similar to what we wrote in our packaging tutorial.
+
+The :code:`Autotools` subclass aims to simplify writing package files and provides
+convenience methods to manipulate each of the different phases for a :code:`Autotools`
+build system.
+
+:code:`Autotools` packages consist of four phases:
+
+1. :code:`autoreconf()`
+2. :code:`configure()`
+3. :code:`build()`
+4. :code:`install()`
+
+
+Each of these phases have sensible defaults. Let's take a quick look at some 
+the internals of the :code:`Autotools` class:
+
+.. note:: 
+    The examples showing code for these classes is abridged to avoid having
+    long examples. We only show what is relevant to the packager.
+
+
+.. literalinclude:: tutorial/examples/Autotools/autotools_class.py
+    :language: python
+    :emphasize-lines: 42,45,62
+    :lines: 40-95,259-267
+    :linenos:
+
+
+Important to note are the highlighted lines. These properties allow the 
```

Comment:
It would probably be good to note in here that the output of `configure_args` is passed to `configure`, so that you shouldn't have to override configure but can instead construct the arguments in `configure_args`.